File: 2018/Day 3/Solution.py
import re, collections
with open('input.txt') as f:
    parse = lambda x: map(int, re.findall(r'\d+', x))
    xs = map(parse, f.read().splitlines())
table = collections.defaultdict(int)
def apply(_, x ,y ,w, h):
    for i in range(y, y+h):
        for j in range(x, x+w):
            table[(i,j)] +=1
def match(lines):
    for (number, x, y, w, h) in lines:
        if sum(table[(i,j)] for i in range(y, y+h)
            for j in range(x, x+w)) == w*h:
                return number
[apply(*line) for line in xs]
overlap = sum([1 for v in table.values() if v > 1])
print('part 1:', overlap)
#print('part 2:', match(xs))
input()

File: 2018/Day 3/Solution.hs
{-#LANGUAGE PartialTypeSignatures, QuasiQuotes#-}
module Solution where

import qualified Data.Map.Lazy as Map
import qualified Data.Set as Set
import Text.RE.TDFA.String
import Text.Printf

type Fabric = Map.Map Int Claim
type Claim = Map.Map Coords Int
type Coords = (Int, Int)

main :: IO()
main = do
  matrix <- Map.fromList.map expand<$>input
  let overlapped = overlap matrix :: Claim
  printf "part 1: %d\n" $ reduce overlapped
  printf "part 2: %d\n" $ isolate overlapped matrix

input :: IO [[Int]]
input = map regex.lines <$> readFile "input.txt"

regex :: Read b => String -> [b]
regex = map read.matches.(*=~ [re|@{%int}|])

expand :: [Int] -> (Int, Claim)
expand [i, x, y, w, h] = (i, Map.fromSet (const 1) lazy_map) where
  lazy_map = Set.fromAscList.map (\[a,b] ->
    (a, b)).sequence $ [[y..y+h-1], [x..x+w-1]]

overlap :: Fabric -> Claim
overlap = Map.unionsWith (+). Map.elems

isolate :: Claim -> Fabric -> Int
isolate master = head. Map.keys. Map.filter match where
  match dict = (== ) dict $Map.intersection master dict

reduce :: Claim -> Int
reduce = foldr qualify 0 where
  qualify x y = if x > 1 then y + 1 else y
